### DISCUSSION TITLE
List all server routes?

### DISCUSSION THREAD

**Discussion by iainsproat** [nuxt/nuxt#20648] [Questions]
Hi,

I'd like to list all server routes - does anyone know how to achieve this?

The reason I would like this is to programatically test all routes, e.g. that they all have appropriate authorization scopes or match OpenAPI spec, or similar functionality that can't be applied via middleware.  I'd prefer not to check each route manually, as this may be subject to human error of missing one or more routes from the tests.

I can see them listed in the generated output, e.g. `.nuxt/zw7297/types/nitro-routes.d.ts`, but how would I get those in vitest or at runtime in the server?

Thanks!

**Comment by huang-julien** [comment]
you can do it through a module i think

**Reply by iainsproat** [reply]
ok, I can see that [a module](https://nuxt.com/docs/guide/going-further/modules#using-hooks) might allow me to [attach to a lifecycle hook](https://nuxt.com/docs/api/advanced/hooks) to get the routes?
I still don't fully know which hook would be best, or what part of the API I would need to call to get the list of routes...

**Reply by huang-julien** [reply]
you can get it with `page:extends`

**Comment by timb-103** [comment]
This would be a great feature to have!

**Comment by maou-shonen** [comment]
in module
```ts
nuxt.hook('ready', () => {
  const nitro = useNitro()
  nitro.hooks.hook('prerender:routes', () => {
    console.log(nitro.scannedHandlers)
  })
})
```

output like...
```ts
{
  handler: '{full_root_path}/server/api/foo.ts',
  lazy: true,
  route: '/api/foo',
  method: undefined
},
{
  handler: '{full_root_path}/server/api/bar.get.ts',
  lazy: true,
  route: '/api/bar',
  method: 'get' 
}
```

Use addTemplate to generate what you need

**Comment by gjedlicska** [comment]
Hey all, thanks for the answers.

@maou-shonen your solution works, my issue is, that this is a build time hook.

I'd like to get the routes at runtime, when the built server starts up.

**Reply by maou-shonen** [reply]
You can use `addTemplate` to generate a json file into #build, and then read it in eventHander.
